File: lib/android/app/src/main/java/com/reactnativenavigation/views/touch/OverlayTouchDelegate.java
package com.reactnativenavigation.views.touch;

import android.graphics.Rect;
import androidx.annotation.VisibleForTesting;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;

import com.reactnativenavigation.parse.params.Bool;
import com.reactnativenavigation.parse.params.NullBool;
import com.reactnativenavigation.utils.StatusBarUtils;
import com.reactnativenavigation.viewcontrollers.IReactView;

public class OverlayTouchDelegate {
    private enum TouchLocation {Outside, Inside}
    private final Rect hitRect = new Rect();
    private IReactView reactView;
    private Bool interceptTouchOutside = new NullBool();

    public void setInterceptTouchOutside(Bool interceptTouchOutside) {
        this.interceptTouchOutside = interceptTouchOutside;
    }

    public OverlayTouchDelegate(IReactView reactView) {
        this.reactView = reactView;
    }

    public boolean onInterceptTouchEvent(MotionEvent event) {
        if (interceptTouchOutside instanceof NullBool) return false;
        switch (event.getActionMasked()) {
            case MotionEvent.ACTION_DOWN:
                return handleDown(event);
            default:
                reactView.dispatchTouchEventToJs(event);
                return false;

        }
    }

    @VisibleForTesting
    public boolean handleDown(MotionEvent event) {
        TouchLocation location = getTouchLocation(event);
        if (location == TouchLocation.Inside) {
            reactView.dispatchTouchEventToJs(event);
            return false;
        }
        return interceptTouchOutside.isFalseOrUndefined();
    }

    private TouchLocation getTouchLocation(MotionEvent ev) {
        View view = getView((ViewGroup) reactView.asView());
        view.getHitRect(hitRect);
        return hitRect.contains((int) ev.getRawX(), (int) ev.getRawY() - StatusBarUtils.getStatusBarHeight(view.getContext())) ?
                TouchLocation.Inside :
                TouchLocation.Outside;
    }

    private View getView(ViewGroup view) {
        return view.getChildCount() > 0 ? view.getChildAt(0) : view;
    }
}

File: lib/android/app/src/main/java/com/reactnativenavigation/views/touch/OverlayTouchDelegate.kt
package com.reactnativenavigation.views.touch

import android.graphics.Rect
import android.view.MotionEvent
import android.view.View
import androidx.annotation.VisibleForTesting
import com.reactnativenavigation.options.params.Bool
import com.reactnativenavigation.options.params.NullBool
import com.reactnativenavigation.viewcontrollers.viewcontroller.IReactView
import com.reactnativenavigation.views.component.ComponentLayout

open class OverlayTouchDelegate(private val component: ComponentLayout, private val reactView: IReactView) {
    var interceptTouchOutside: Bool = NullBool()
    private val hitRect = Rect()
    private val overlayView: View
        get() = if (reactView.asView().childCount > 0) reactView.asView().getChildAt(0) else reactView.asView()

    fun onInterceptTouchEvent(event: MotionEvent): Boolean {
        return when (interceptTouchOutside.hasValue() && event.actionMasked == MotionEvent.ACTION_DOWN) {
            true -> handleDown(event)
            false -> component.superOnInterceptTouchEvent(event)
        }
    }

    @VisibleForTesting
    open fun handleDown(event: MotionEvent) = if (isTouchInsideOverlay(event)) component.superOnInterceptTouchEvent(event) else interceptTouchOutside.isFalse

    private fun isTouchInsideOverlay(ev: MotionEvent): Boolean {
        overlayView.getHitRect(hitRect)
        return hitRect.contains(ev.rawX.toInt(), (ev.rawY - component.y).toInt())
    }
}